ENH: Implement reshape method on sparse matrices

drhagen opened the pull request:
Currently, the reshape method is defined on the base class of sparse matrices, but only implemented for LIL matrices. This defines an implementation for COO sparse matrices and then adds a implementation to the base class that first converts to COO to use its reshape then converts back.

The discussion began at #3507. The original COO implementation is merged from #3957.

jnothman commented on scipy/sparse/coo.py:
```diff
@@ -190,6 +190,31 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
 
         self._check()
 
+    def reshape(self, shape, order='C', copy=False):
```
You should also allow the *args shape specification

drhagen replied:
I don't feel strongly either way, but I am disinclined to add it until Numpy does it first to either the `reshape` function or `reshape` method on `ndarray`.

jnothman replied:
But it has that functionality... X.reshape(5,6)

drhagen replied:
Huh, you are correct. That's apparently undocumented. I am not even sure what the exact behavior is.

drhagen replied:
It appears that `ndarray.reshape` first checks if there is more than one positional argument; if there is, it tries to cast all positional arguments to ints and joins them into a tuple. If there is only one argument, it tries to iterate over it and cast all contents to ints and joins them into a tuple. If it succeeds in iterating, but fails at casting it raises the casting error. If iterating fails, it then tries to cast sole argument to an int and puts it in a tuple by itself.

The `shape` parameter is a positional-only argument. The `order` parameter is a named-only argument. This suggests the true signature is `reshape(shape*, order='C')`

Is there a function in Numpy somewhere that does this shape args unification so I can be sure I am consistent?

jnothman commented on scipy/sparse/lil.py:
```diff
@@ -142,6 +142,11 @@ def set_shape(self,shape):
             raise ValueError('invalid shape')
 
         if (self._shape != shape) and (self._shape is not None):
+            raise NotImplementedError(
```
This will, I presume, require a deprecation period.

drhagen replied:
That was left over from the previous PR, presumably because the existing implementation was completely broken. I can't think of a reason to actually block setting the shape in place, so why don't I just implement the `shape` property for everything.

perimosocordiae commented:
I'm -1 on removing support for reshaping by assigning to the `.shape` attribute. It's not pretty, but it matches numpy and it's been in use for years now.

I'm also -1 on all the current changes to `lil.py`:
 - The updated version of `tocsr` is essentially the same as the current implementation.
 - An efficient implementation of LIL reshape is definitely going to be faster than LIL -> tocoo -> reshape -> tolil. We should just improve the existing method.

That said, I'm glad you're taking this on, @drhagen ! It has been a long time, and it's still a useful enhancement.

drhagen commented:
> I'm -1 on removing support for reshaping by assigning to the .shape attribute.

Technically, this isn't removing functionality. Look at the implementation in LIL (assigns to `self`); it does literally nothing and no one noticed for years. But I am planning to implement assignment to `shape` anyway to maintain compatibility with Numpy.

> The updated version of `tocsr` is essentially the same as the current implementation.

I guess I don't understand the -1 for code that's "essentially the same as the current implementation" at a third the lines.

> An efficient implementation of LIL reshape is definitely going to be faster than LIL -> tocoo -> reshape -> tolil. We should just improve the existing method.

I'll wait to do some benchmarks before deciding which technique to ultimately keep.

perimosocordiae commented on scipy/sparse/base.py:
```diff
@@ -125,8 +116,8 @@ def reshape(self, shape, order='C'):
         """
-        raise NotImplementedError("Reshaping not implemented for %s." %
-                                  self.__class__.__name__)
+        return (self.tocoo(copy=copy).reshape(*args, order=order, copy=True)
```
Why `copy=True` here?

drhagen replied:
That should be `False`.

perimosocordiae commented on scipy/sparse/base.py:
```diff
@@ -125,8 +116,8 @@ def reshape(self, shape, order='C'):
-        raise NotImplementedError("Reshaping not implemented for %s." %
-                                  self.__class__.__name__)
+        return (self.tocoo(copy=copy).reshape(*args, order=order, copy=True)
+                .asformat(self.format))
```
Some existing methods have a `format=None` keyword argument, where `None` means to return the most efficient / convenient format. Might be useful here.

drhagen replied:
I'm not opposed to this idea, but I couldn't find an example of the format option in `base.py`. Maybe we can defer this to a larger refactoring for including this option.

I did find that converting back to the same format is not common, however. The default for most methods is `return self.tocsr().some_operation()`. I think I'll try to write a `reshape` method for CSR and use that as the default instead. If users got back a different type from `reshape`, I think they would be OK with a CSR, but probably not a COO.

drhagen replied:
The more I looked at it, the COO matrix is a first class citizen of the sparse package and a perfectly fine object to return. I looked at the CSC reshape implementation in the Octave source, but couldn't figure out how to implement that in Python that would likely be faster than just using the existing C code to cast to COO and back. If someone knows how to do C work, they could probably copy the algorithm from Octave and wire it into `_sparsetools`.

ewmoore replied:
Please don't look at the Octave source if you intend to contribute similar things here.  Octave is GPL and Scipy is not compatible with that license.

drhagenabm replied:
That is a good point. I guess I'll leave the clean room implementation of CSR reshape to someone else.

perimosocordiae commented on scipy/sparse/bsr.py:
```diff
@@ -13,7 +13,8 @@
 from .base import isspmatrix, _formats, spmatrix
-from .sputils import isshape, getdtype, to_native, upcast, get_index_dtype
+from .sputils import isshape, getdtype, to_native, upcast, get_index_dtype, \
+    check_shape
```
Use parentheses in favor of line continuations.

drhagen replied:
I'll fix it.

perimosocordiae commented on scipy/sparse/coo.py:
```diff
@@ -190,6 +190,25 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
+            flat_indices = self.row + nrows * self.col
+            new_col, new_row = divmod(flat_indices, shape[0])
+        else:
+            ValueError("'order' must be 'C' or 'F'")
```
Forgot the `raise` here.

drhagen replied:
Yep...

perimosocordiae commented on scipy/sparse/coo.py:
```diff
@@ -190,6 +190,25 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
+            ValueError("'order' must be 'C' or 'F'")
+
+        return coo_matrix((self.data, (new_row, new_col)),
+                          shape=shape, copy=copy)
```
The only thing that needs copying if `copy=True` is `self.data`, so I'd handle that here and pass `copy=False` to `coo_matrix`. Otherwise, you're making extra copies of `new_row` and `new_col`.

drhagen replied:
This is a good idea, and something to keep in mind if more `reshape` methods get specialized.

perimosocordiae commented on scipy/sparse/sputils.py:
```diff
@@ -251,6 +251,61 @@ def validateaxis(axis):
+        if isinstance(arg, int_types):
+            return int(arg)
+        elif isinstance(arg, np.ndarray) and np.isscalar(arg) and \
+                arg.dtype in int_types:
```
`np.issubdtype(arg.dtype, np.integer)`

drhagen replied:
I wish I knew about this ages ago.

perimosocordiae commented on scipy/sparse/sputils.py:
```diff
@@ -251,6 +251,61 @@ def validateaxis(axis):
+                            "integer".format(args[0].__class__.__name__))
+
+    if len(args) == 0:
+        raise TypeError("'reshape' missing 1 required positional argument: "
```
This function isn't always called from `reshape`, so this error message could be misleading.

drhagen replied:
Technically, only the `reshape` function can have zero args, but we might want to reuse this function for other shape checks, so I'll make it more generic.

perimosocordiae commented on scipy/sparse/sputils.py:
```diff
@@ -251,6 +251,61 @@ def validateaxis(axis):
+
+    # Check for negatives
+    if new_shape[0] < 0 or new_shape[1] < 0:
+        raise ValueError("'shape' elements cannot be negative")
```
We should support a single `-1` dimension: `foo.reshape((5, -1))`

drhagen replied:
I forgot about this case. I'll have to change some of the sparse tests, because they expect an error.

drhagen commented:
I've got the changes in that I want to make. I added the ability to have one index be negative like Numpy allows. I resurrected the LIL-specific implementation and expanded it for `order='F'`. I also changed the default implementation to simply return the COO matrix rather than reformat it as the current class because returning a sparse matrix in any format is the current standard behavior of the module. There are a handful of bugfixes as recommended by @perimosocordiae.

perimosocordiae commented:
The test failure on Python 2 is real:
```
E     File "/home/travis/build/scipy/scipy/build/testenv/lib/python2.7/site-packages/scipy/sparse/base.py", line 91
E       def reshape(self, *args, order='C', copy=False):
E                                    ^
E   SyntaxError: invalid syntax
```

Unfortunately I think the only way around this is to use `**kwargs`.

drhagen commented:
> The test failure on Python 2 is real

How does Scipy deal with this normally? There is a [library](https://pypi.python.org/pypi/kwonly-args) that looks like it handles this, but it does not appear to be a dependency of Scipy.

jnothman commented:
Just use **kwargs, pop elements off and assert that it is empty at the end.

drhagen commented:
@jnothman What's the Scipy policy on testing intentionally undocumented behavior?

jnothman commented:
NFI

perimosocordiae commented:
I'd appreciate the test case, even if it's not documented. User code breaking is something the test suite should tell us about.

perimosocordiae commented on scipy/sparse/base.py:
```diff
@@ -77,56 +78,58 @@ def __init__(self, maxprint=MAXPRINT):
-        raise NotImplementedError("Reshaping not implemented for %s." %
-                                  self.__class__.__name__)
+        # If the shape already matches, don't bother doing an actual reshape
+        # Otherwise, the default is ot convert to COO and use its reshape
```
typo: to

perimosocordiae commented on scipy/sparse/sputils.py:
```diff
@@ -251,6 +251,88 @@ def validateaxis(axis):
 
+def check_shape(args, current_shape=None):
+    """Imitate numpy.matrix handling of shape arguments"""
+    def check_int(arg):
```
I think you can replace this function with `operator.index(arg)`. (See `sputils.isintlike`.)

drhagen replied:
Yep, `operator.index` what I was looking for. It seems to be completely consistent with how `numpy.matrix` handles them, like I intended. I see that you recently updated `sputils.isinitlike` to use it.

drhagen commented:
Ok, I've made the final edits and ran the tests. If there isn't anything more, I think this is ready to be merged.

perimosocordiae commented:
Ok, I'll wait a few days to see if anyone else has comments, then merge.

drhagen commented:
Are we good to merge?

perimosocordiae commented:
Yep, merged. Thanks, @drhagen !
